source/val/ValidationState.cpp

```diff
@@ -357,4 +365,91 @@ spv_result_t ValidationState_t::RegisterFunctionEnd() {
   return SPV_SUCCESS;
 }
 
+ValidationState_t::UseDefTracker::UseDefTracker(ValidationState_t* state)
+    : module_(state) {}
+
+void ValidationState_t::UseDefTracker::AddDef(const spv_id_info_t& def) {
+  all_defs_[def.id] = def;
+  if (module_->in_block()) {
+    auto block = module_->get_current_function()->get_current_block();
+    block_defs_[block->get_id()].push_back(def.id);
+  } else {
+    module_defs_.insert(def.id);
+  }
+}
+
+void ValidationState_t::UseDefTracker::AddUse(uint32_t id, bool is_phi) {
+  all_uses_.insert(id);
+  if (module_->in_block()) {
+    auto block = module_->get_current_function()->get_current_block();
+    block_uses_[block->get_id()].insert({id, is_phi});
+  }
+}
+
+std::pair<bool, spv_id_info_t> ValidationState_t::UseDefTracker::FindDef(
+    uint32_t id) const {
+  if (all_defs_.count(id) == 0) {
+    return std::make_pair(false, spv_id_info_t{});
+  } else {
+    /// We are in a const function, so we cannot use defs.operator[]().
+    /// Luckily we know the key exists, so defs_.at() won't throw an
+    /// exception.
+    return std::make_pair(true, all_defs_.at(id));
+  }
+}
+
+std::unordered_set<uint32_t>
+ValidationState_t::UseDefTracker::FindUsesWithoutDefs() const {
+  auto diff = all_uses_;
+  for (const auto d : all_defs_) diff.erase(d.first);
+  return diff;
+}
+
+void ValidationState_t::UseDefTracker::UpdateDefs(Function* func) {
+  auto& ordered_blocks = func->get_blocks();
+
+  // Because dominators must appear before the blocks they dominate in the
+  // binary , we only need to visit the immidiate dominators to propagate the
+  // ids throughout the CFG
+  for (auto block : ordered_blocks) {
+    auto& block_defs = block_defs_[block->get_id()];
+    auto idom = block->GetImmediateDominator();
+    if (block->is_reachable() && block != idom) {
+      assert(idom);
+      auto& idom_defs = block_defs_[idom->get_id()];
+      block_defs.reserve(block_defs.size() + idom_defs.size());
+      copy(begin(idom_defs), end(idom_defs), back_inserter(block_defs));
+    } else {
+      // This is the first block or it is unreachable in the CFG
+      block_defs.reserve(block_defs.size() + module_defs_.size());
+      copy(begin(module_defs_), end(module_defs_), back_inserter(block_defs));
+    }
+    sort(begin(block_defs), end(block_defs));
+  }
+}
+
+std::map<uint32_t, vector<uint32_t>>
+ValidationState_t::UseDefTracker::CheckDefs(Function* func) {
+  std::map<uint32_t, vector<uint32_t>> undef_ids;
+
+  for (auto block : func->get_blocks()) {
+    auto& block_defs = block_defs_[block->get_id()];
+    auto& block_uses = block_uses_[block->get_id()];
+    vector<uint32_t> uses_vec;
+    uses_vec.reserve(block_uses.size());
+    for (auto& use : block_uses) {
+      if (use.second == false) {
+        uses_vec.push_back(use.first);
```

Comment:
Ah.  I think label IDs are not counted as definitions, since the validation state is only "in a block" *after* the label has been fully processed.  That's a weird accident because the ProcessIds handler is called before CfgPass.
And so the "Current block" is not registered at the moment you see the label.  So then the label ID is put into the module scope bucket.  That's *really* confusing.  For one thing, the block_defs_ vectors will all be bloated without good reason, and the module_defs_ is incorrectly named.

I think we need to be more careful about partitioning the data.